esp-hal/src/timer/systimer.rs

```diff
@@ -138,150 +152,551 @@ impl<'d> SystemTimer<'d, Blocking> {
+pub enum ComparatorMode {
+    /// The comparator will generate interrupts periodically.
+    Period,
+
+    /// The comparator will generate an interrupt when the unit reaches the
+    /// target.
+    Target,
+}
+
+impl Unit<'static, 0> {
+    /// Conjure a system timer unit out of thin air.
+    ///
+    /// # Safety
+    ///
+    /// Users must take care to ensure that only one reference to the unit is
+    /// in scope at any given time.
+    pub const unsafe fn conjure() -> Self {
+        Self(PhantomData)
+    }
+}
+
+#[cfg(not(esp32s2))]
+impl Unit<'static, 1> {
+    /// Conjure a system timer unit out of thin air.
+    ///
+    /// # Safety
+    ///
+    /// Users must take care to ensure that only one reference to the unit is
+    /// in scope at any given time.
+    pub const unsafe fn conjure() -> Self {
+        Self(PhantomData)
     }
 }
 
-impl<T, const CHANNEL: u8> InterruptConfigurable for Alarm<T, Blocking, CHANNEL> {
+/// A unit whose value cannot be updated.
+pub struct FrozenUnit<'d, const CHANNEL: u8>(&'d Unit<'d, CHANNEL>);
+
+impl<'d, const CHANNEL: u8> FrozenUnit<'d, CHANNEL> {
+    /// Creates a frozen unit. You will no longer be allowed
+    /// direct access to this unit until all the alarms created
+    /// from the unit are dropped.
+    pub fn new(unit: &'d mut Unit<'d, CHANNEL>) -> Self {
+        Self(unit)
+    }
+
+    fn borrow(&self) -> &'d Unit<'d, CHANNEL> {
+        self.0
+    }
+}
+
+/// Alarms created from the System Timer peripheral.
+pub struct SysTimerAlarms<MODE, DM: Mode> {
+    /// Alarm 0
+    pub alarm0: Alarm<'static, MODE, DM, 0, 0>,
+    /// Alarm 1
+    pub alarm1: Alarm<'static, MODE, DM, 1, 0>,
+    /// Alarm 2
+    pub alarm2: Alarm<'static, MODE, DM, 2, 0>,
+
+    /// Unit 1
+    ///
+    /// Leftover unit which wasn't used to create the three alarms.
+    #[cfg(not(esp32s2))]
+    pub unit1: Unit<'static, 1>,
+}
+
+/// A marker for a [Alarm] in target mode.
+#[derive(Debug)]
+pub struct Target;
+
+/// A marker for a [Alarm] in periodic mode.
+#[derive(Debug)]
+pub struct Periodic;
+
+/// A single alarm.
+pub struct Alarm<'d, MODE, DM, const COMP: u8, const UNIT: u8>
+where
+    DM: Mode,
+{
+    comparator: Comparator<'d, COMP>,
+    unit: &'d Unit<'d, UNIT>,
+    config: &'d Mutex<Cell<Option<Config<'d>>>>,
+    _pd: PhantomData<(MODE, DM)>,
+}
+
+impl<'d, T, DM, const COMP: u8, const UNIT: u8> Debug for Alarm<'d, T, DM, COMP, UNIT>
+where
+    DM: Mode,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Alarm")
+            .field("comparator", &COMP)
+            .field("unit", &UNIT)
+            .finish()
+    }
+}
+
+impl<'d, T, const COMP: u8, const UNIT: u8> Alarm<'d, T, Blocking, COMP, UNIT> {
+    /// Creates a new alarm from a comparator and unit, in blocking mode.
+    pub fn new(
+        comparator: Comparator<'d, COMP>,
+        unit: &FrozenUnit<'d, UNIT>,
+        config: &'d Mutex<Cell<Option<Config<'d>>>>,
```

Comment:
I'm not a fan of the config way of doing this, especially since the config is a zst, so we're adding all this complexity just to add some synchronization. If I understand you in https://github.com/esp-rs/esp-hal/issues/1477#issuecomment-2184010689, can't we just use a `critical_section` when modifying these registers?

Reply:
I go into it more in https://github.com/esp-rs/esp-hal/issues/1874#issuecomment-2255784263, but `critical_section` is a sledge hammer of a synchronization tool. Basically a global mutex, and using a global mutex to protect this one register feels inappropriate. This should really have a "local" mutex, that doesn't unnecessarily prevent the other core from locking something unrelated. This "local" mutex basically just means have a spinlock per register, and the `Config` object just provides something for users to prove mutual exclusion.

In my opinion, users should be free to pick any (or no) synchronization primitive for their use case. Not everyone will want to disable interrupts or block the other core from running. The `Config` object seemed like the way to achieve this (though I can admit it's not the prettiest).

Reply:
Unfortunately because `Alarm`s are stored in `'static` and "impl Trait in type alias"/"impl Trait in static" isn't stable yet. I'm forced to hard code a Mutex as the synchronization primitive for `Alarm`s.

Reply:
Ah sorry I missed that comment. I do generally agree with you regarding `critical_section::with` being very heavy, though I do wonder the real-world impact in this case. Also, it's currently using a `critical_section::Mutex` which forces `critical_section` anyways.

I would prefer the simpler API over absolute perf, unless we can demonstrate that the performance is absolutely required for normal operation of the driver.

Reply:
> I was worried you might say that😅

Sorry :sweat_smile:, but I think for now yes, we should remove `Config` and just use `critical_section` internally.